code/modules/projectiles/guns/projectile/shotgun.dm

```diff
@@ -2,6 +2,43 @@
 /obj/item/weapon/gun/projectile/shotgun
 	icon = 'icons/urist/items/shotguns.dmi'
 
+/obj/item/weapon/gun/projectile/shotgun/load_ammo(var/obj/item/A, mob/user)
+	if(istype(A, /obj/item/ammo_magazine))
+		. = TRUE
+		var/obj/item/ammo_magazine/AM = A
+		if(caliber != AM.caliber)
+			return //incompatible
+
+		if(AM.mag_type == SPEEDLOADER)
+			if(loaded.len >= max_shells)
+				to_chat(user, "<span class='warning'>[src] is full!</span>")
+				return
+			var/obj/item/ammo_casing/C = AM.stored_ammo[AM.stored_ammo.len]
+			AM.stored_ammo-=C
+			loaded.Insert(1, C) //add to the head of the list
+			user.visible_message("[user] inserts \a [C] into [src].", "<span class='notice'>You insert \a [C] into [src].</span>")
+			playsound(loc, load_sound, 50, 1)
+			AM.update_icon()
+			if(istype(AM, /obj/item/ammo_magazine/shotbundle))
+				if(!AM.stored_ammo.len)
+					if(!user.unEquip(AM, src))
+						return
+					AM.update_icon()
+	else if(istype(A, /obj/item/ammo_casing))
+		. = TRUE
+		var/obj/item/ammo_casing/C = A
+		if(!(load_method & SINGLE_CASING) || caliber != C.caliber)
+			return //incompatible
+		if(loaded.len >= max_shells)
+			to_chat(user, "<span class='warning'>[src] is full.</span>")
+			return
+		if(!user.unEquip(C, src))
+			return
+		loaded.Insert(1, C) //add to the head of the list
+		user.visible_message("[user] inserts \a [C] into [src].", "<span class='notice'>You insert \a [C] into [src].</span>")
+		playsound(loc, load_sound, 50, 1)
+	update_icon()
+
```

Comment:
```suggestion
/obj/item/weapon/gun/projectile/shotgun/load_ammo(var/obj/item/A, mob/user)
	if(istype(A, /obj/item/ammo_magazine/shotbundle))
		. = TRUE
		var/obj/item/ammo_magazine/shotbundle/SB = A
		if(caliber != SB.caliber)
			return

		if(loaded.len >= max_shells)
			to_chat(user, "<span class='warning'>[src] is full!</span>")
			return

		var/obj/item/ammo_casing/C = SB.stored_ammo[SB.stored_ammo.len]
		SB.stored_ammo -= C
		loaded.Insert(1, C)
		user.visible_message("[user] inserts \a [C] into [src].", "<span class='notice'>You insert \a [C] into [src].</span>")
		playsound(loc, load_sound, 50, 1)
		SB.update_icon()
		SB.check_ammo_count()
		return

	return ..()
```

Reply:
This removes the one-by-one for shell holders. You should honestly just mark all shotguns as accepting speedloaders and overwrite the attack_self behavior for shotguns to set allow_dump=0 save for the double barrel.